Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift

```diff
@@ -0,0 +1,122 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Async Algorithms open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+extension AsyncSequence where Self: Sendable {
+  /// Creates an asynchronous sequence that buffers elements.
+  ///
+  /// The buffering behaviour is dictated by the policy:
+  /// - bounded: will buffer elements until the limit is reached. Then it will suspend the upstream async sequence.
+  /// - unbounded: will buffer elements without limit.
+  /// - bufferingLatest: will buffer elements until the limit is reached. Then it will discard the oldest elements.
+  /// - bufferingOldest: will buffer elements until the limit is reached. Then it will discard the newest elements.
+  ///
+  /// - Parameter policy: A policy that drives the behaviour of the ``AsyncBufferSequence``
+  /// - Returns: An asynchronous sequence that buffers elements up to a given limit.
+  public func buffer(
+    policy: AsyncBufferSequence<Self>.Policy
+  ) -> AsyncBufferSequence<Self> {
+    AsyncBufferSequence<Self>(base: self, policy: policy)
+  }
+}
+
+/// An `AsyncSequence` that buffers elements in regard to a policy.
+public struct AsyncBufferSequence<Base: AsyncSequence & Sendable>: AsyncSequence {
+  /// A policy dictating the buffering behaviour of an ``AsyncBufferSequence``
+  public struct Policy: Sendable {
+    let policy: _Policy
+
+    /// A policy for buffering elements until the limit is reached.
+    /// Then consumption of the upstream `AsyncSequence` will be paused until elements are consumed from the buffer.
+    public static func bounded(_ limit: Int) -> Self {
+      precondition(limit > 0, "The limit should be positive for the buffer operator to be efficient.")
```

Comment:
What happens if limit is 0? There are scenarios where that could be useful.

Reply:
Interesting question. Before taking any actions I think we have to think about that.

The current behaviour will be:
- bounded policy: after the first call to next, the consumer will receive the first element and then the producer will suspend until the consumer request a next element. In the end it's like the buffer iterator has no effect here.

- bufferingNewest: after the first call to next, the consumer will receive the first element and then the producer will continue to emit elements. The buffer will stack only the latest emitted element (the previous one is discarded). When a new call to next is done then the penultimate element is consumed.

- bufferingOldest: after the first call to next, the consumer will receive the first element and then the producer will continue to emit elements. The buffer will never stack elements. We loose all the elements until the consumer request a next element.

I guess this is not ideal for the two last cases. Perhaps we should also suspend the producer in those cases?

What do you think @FranzBusch @phausler ?

Reply:
@phausler @FranzBusch 

I've tried something. Considering that a limit <=0 should be equivalent to having no buffer, I've introduced a third `storageType` in the enum. This new case is `transparent` and just forwards the call to next to the upstream async sequence, as if there was no buffer at all. 

I've added some unit tests to make sure the upstream sequence timeline is not altered. I've also added comments in the doc and removed the preconditions.

Let me know what you think.

Reply:
I think handling `0` with no buffer is fine; however, I still think we should `preconditionFailure` on `<0`.

Reply:
done.